### PR TITLE
An incoming offer can generate simulcast

### PR DESCRIPTION
**Pull request by alvestrand**
This test needs discussion - my reading of draft-ietf-mmusic-simulcast is that an incoming offer with two RIDs should cause a transceiver to be created with two encodings avaialble, and an answer with two rid lines with send direction, indicating that the answerer is capable of sending 2 layers of simulcast.

This currently fails on both Chrome and Firefox.

### CODE REVIEW

**Comment by alvestrand** [comment]
Actually JSEP says:
```
"  In addition, JSEP does not provide a mechanism to handle an incoming
   offer requesting simulcast from the JSEP endpoint.  This means that
   setting up simulcast in the case where the JSEP endpoint receives the
   initial offer requires out-of-band signaling or SDP inspection.
   However, in the case where the JSEP endpoint sets up simulcast in its
   in initial offer, any established simulcast streams will continue to
   work upon receipt of an incoming re-offer.  Future versions of this
   specification may add additional APIs to handle the incoming initial
   offer scenario.
"
```
The SDP in the test would be reasonable if it was supposed to work, but apparently it's not supposed to.

**Comment by nils-ohlmeier** [comment]
In Firefox this should work, but you need to set the Encodings before setting the Offer.

**Comment by fippo** [comment]
You might want to call getUserMedia, otherwise the m-line will be inactive...

http://jsfiddle.net/xcmq4vbn/1/ works in Firefox and has less boilerplate than the Firefox test. Calling setParameters works after setting the offer.

File: webrtc/protocol/simulcast-negotiate.html [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+a=simulcast:recv 1;2
+a=rid:1 recv max-fps=60
+a=rid:2 recv max-fps=30
+`;
+      const pc = new RTCPeerConnection();
+      await pc.setRemoteDescription({type: 'offer', sdp: offer_sdp});
+      const answer = await pc.createAnswer();
+      const rid_lines = answer.sdp.split('\r\n')
+            .filter(line => line.match('^a=rid'));
+      assert_equals(rid_lines.length, 2, 'Answer shold have 2 RIDs');
```
Comment (youennf) [inline webrtc/protocol/simulcast-negotiate.html]: s/shold/should/

File: webrtc/protocol/simulcast-negotiate.html [hunk 2]
```diff
@@ -0,0 +1,50 @@
+<!doctype html>
+<meta charset=utf-8>
+<title>Negotiate Simulcast</title>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="../RTCPeerConnection-helper.js"></script>
```
Comment (youennf) [inline webrtc/protocol/simulcast-negotiate.html]: Is it used?

File: webrtc/protocol/simulcast-negotiate.html [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+<script>
+'use strict';
+
+
+
```
Comment (youennf) [inline webrtc/protocol/simulcast-negotiate.html]: Two unneeded empty lines.

**Review by youennf — COMMENTED** [review]
> In Firefox this should work, but you need to set the Encodings before setting the Offer.
> 
> Test code https://searchfox.org/mozilla-central/source/dom/media/tests/mochitest/test_peerConnection_simulcastAnswer.html#59

In such a case, do you need to have simulcast enabled in the offer?

**Comment by nils-ohlmeier** [comment]
> In such a case, do you need to have simulcast enabled in the offer?

I'm not sure I get what you mean. Yes the offer needs to contain a=simulcast and a=rid attributes to be able to turn on simulcast in the Answer.

**Comment by aboba** [comment]
> Yes the offer needs to contain a=simulcast and a=rid attributes to be able to turn on simulcast

[BA] Yes, but other than the number of encodings and the rids are there expectations that SDP attributes in the test's Offer (max-fr) will be translated into encoding parameters on the Answering side when the transceiver is evented?

File: webrtc/protocol/simulcast-negotiate.html (hunk 1 above)
Comment (aboba) [reply webrtc/protocol/simulcast-negotiate.html]: Are we expecting the max-fr attributes to translate into maxFramerate values within the encoding parameters?

Comment (alvestrand) [reply webrtc/protocol/simulcast-negotiate.html]: If they have any effect at all, that's the effect that they should have.

File: webrtc/protocol/simulcast-negotiate.html (hunk 2 above)
Comment (alvestrand) [reply webrtc/protocol/simulcast-negotiate.html]: Not yet.

**Comment by alvestrand** [comment]
The number of encodings is really all I was looking for with this test. Nils, you're saying that you can do
<pre>
pc.setRemoteDescription(offer);
pc.getSenders().setParameters(.... some simulcast stuff ...);
</pre>
and get simulcast, right?
According to spec, this should be pc.getSenders().setParameters(modifyParameters(pc.getSenders().getParameters())), but mozilla doesn't check this - Chrome does.
Do you have getParameters() at all at the moment?

**Comment by alvestrand** [comment]
(I see that getParameters on Firefox ESR returns a rather empty thing. Has it improved?)

**Comment by youennf** [comment]
> > In such a case, do you need to have simulcast enabled in the offer?
> 
> I'm not sure I get what you mean. Yes the offer needs to contain a=simulcast and a=rid attributes to be able to turn on simulcast in the Answer.

I was wondering whether just setParameters in Firefox would trigger simulcast (not as per spec but hey...). From your answer, it seems not.

My main question is to understand whether this is nice to have or required to support SDP simulcast negotiation, especially with the trend to move out of SDP negotiation.

It seems a SFU needs to send some simulcast parameters to the web page so that setParameters is called.
The SFU could instead send the number of participants to let the web page create an offer with a sendonly simulcast section and the necessary recvonly sections.
It would then receive an offer that matches what it wants. This is racy but no more than the alternative.
What are we loosing in doing that?

**Comment by fippo** [comment]
> The SFU could instead send the number of participants to let the web page create an offer with a sendonly simulcast section and the necessary recvonly sections.

This directly affects things like ICE and DTLS roles and also changes behaviour in weird ways, see e.g. [here](https://bugs.chromium.org/p/webrtc/issues/detail?id=9931) for an example.

**Comment by alvestrand** [comment]
Status: Needs a spec update to make sense.

**Comment by foolip** [comment]
This PR is blocked on the required "Travis CI - Pull Request" check after https://github.com/web-platform-tests/wpt/issues/14499. In order to trigger it, I will close and reopen this PR.

**Comment by alvestrand** [comment]
I think the spec has caught up to this one now.
@youennf can you take another look?

**Comment by alvestrand** [comment]
Closing in favor of a PR submitted via the chromium pipeline.

**Comment by alvestrand** [comment]
Chromium review: https://chromium-review.googlesource.com/c/chromium/src/+/1540306

**Comment by foolip** [comment]
Back on July 2 I got a reply from GitHub support about this:

> We were wondering if you can try removing 14 labels from the PR and giving the close another go? There's a maximum label limit our engineering team think is causing the PR not to close.
>
> That limitation is a recent addition, which means it won't be possible to apply that many labels to a PR going forward. It seems this change may have some unintended side effects for older PRs however.

I'll try that now.

**Pull request closed without merging**